extensions/notebook/src/book/bookModel.ts

```diff
@@ -0,0 +1,221 @@
+	private _tableOfContentPaths: string[] = [];
+	readonly providerId: string = 'BookNavigator';
+	// For testing
+	private _errorMessage: string;
+
+	constructor(bookPath: string, openAsUntitled: boolean, extensionContext: vscode.ExtensionContext) {
+		this.BookPath = bookPath;
+		this.openAsUntitled = openAsUntitled;
+		this._extensionContext = extensionContext;
+		extensionContext.subscriptions.push(azdata.nb.registerNavigationProvider(this));
+		//this.getTableOfContentFiles(this.BookPath).then(() => { this._books = this.readBooks();}, (err) => { console.log(err); });
+
+	}
+
+	public async initializeContents(): Promise<void> {
+		await this.getTableOfContentFiles(this.BookPath);
+		this.readBooks();
+		return;
+	}
+
+	public getAllBooks(): Map<string, BookTreeItem> {
+		return this._allNotebooks;
+	}
+
+	async getTableOfContentFiles(workspacePath: string): Promise<void> {
+		try {
+			let notebookConfig = vscode.workspace.getConfiguration(notebookConfigKey);
+			let maxDepth = notebookConfig[maxBookSearchDepth];
+			// Use default value if user enters an invalid value
+			if (maxDepth === undefined || maxDepth < 0) {
+				maxDepth = 5;
+			} else if (maxDepth === 0) { // No limit of search depth if user enters 0
+				maxDepth = undefined;
+			}
+
+			let p = path.join(workspacePath, '**', '_data', 'toc.yml').replace(/\\/g, '/');
+			let tableOfContentPaths = await glob(p, { deep: maxDepth });
+			this._tableOfContentPaths = this._tableOfContentPaths.concat(tableOfContentPaths);
+			let bookOpened: boolean = this._tableOfContentPaths.length > 0;
+			vscode.commands.executeCommand('setContext', 'bookOpened', bookOpened);
+		} catch (error) {
+			console.log(error);
+		}
+	}
+
+	public readBooks(): BookTreeItem[] {
+		for (const contentPath of this._tableOfContentPaths) {
+			let root = path.dirname(path.dirname(contentPath));
+			try {
+				const config = yaml.safeLoad(fs.readFileSync(path.join(root, '_config.yml'), 'utf-8'));
+				const tableOfContents = yaml.safeLoad(fs.readFileSync(contentPath, 'utf-8'));
+				let book = new BookTreeItem({
+					title: config.title,
+					root: root,
+					tableOfContents: { sections: this.parseJupyterSection(tableOfContents) },
+					page: tableOfContents,
+					type: BookTreeItemType.Book,
+					treeItemCollapsibleState: vscode.TreeItemCollapsibleState.Expanded,
+					untitled: this.openAsUntitled,
+				},
+					{
+						light: this._extensionContext.asAbsolutePath('resources/light/book.svg'),
+						dark: this._extensionContext.asAbsolutePath('resources/dark/book_inverse.svg')
+					}
+				);
+				this._bookItems.push(book);
+			} catch (e) {
+				let error = e instanceof Error ? e.message : e;
+				this._errorMessage = error;
+				vscode.window.showErrorMessage(error);
+			}
+		}
+		return this._bookItems;
+	}
+
+	public get getBookItems() {
+		return this._bookItems;
+	}
+
+	public getSections(tableOfContents: IJupyterBookToc, sections: IJupyterBookSection[], root: string): BookTreeItem[] {
```

Comment:
Can this be an async function - so you can use fs.exists() instead of fs.existsSync()